sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/TableCatalog.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalog.v2;
+
+import org.apache.spark.sql.catalog.v2.expressions.Transform;
+import org.apache.spark.sql.catalyst.analysis.NoSuchNamespaceException;
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
+import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException;
+import org.apache.spark.sql.sources.v2.Table;
+import org.apache.spark.sql.types.StructType;
+
+import java.util.Map;
+
+/**
+ * Catalog methods for working with Tables.
+ */
+public interface TableCatalog extends CatalogPlugin {
+  /**
+   * List the tables in a namespace from the catalog.
+   * <p>
+   * If the catalog supports views, this must return identifiers for only tables and not views.
+   *
+   * @param namespace a multi-part namespace
+   * @return an array of Identifiers for tables
+   * @throws NoSuchNamespaceException If the namespace does not exist (optional).
+   */
+  Identifier[] listTables(String[] namespace) throws NoSuchNamespaceException;
+
+  /**
+   * Load table metadata by {@link Identifier identifier} from the catalog.
+   * <p>
+   * If the catalog supports views and contains a view for the identifier and not a table, this
+   * must throw {@link NoSuchTableException}.
+   *
+   * @param ident a table identifier
+   * @return the table's metadata
+   * @throws NoSuchTableException If the table doesn't exist or is a view
+   */
+  Table loadTable(Identifier ident) throws NoSuchTableException;
+
+  /**
+   * Invalidate cached table metadata for an {@link Identifier identifier}.
+   * <p>
+   * If the table is already loaded or cached, drop cached data. If the table does not exist or is
+   * not cached, do nothing. Calling this method should not query remote services.
+   *
+   * @param ident a table identifier
+   */
+  default void invalidateTable(Identifier ident) {
```

Comment:
what's the use case of this method? I can only think of one use case: when end users run `REFRESH TABLE`. Is there any more use cases?

Reply:
This will be called by `REFRESH TABLE` to invalidate cached data. This was previously named `refreshTable`, but you preferred a void return type. In that case, `invalidateTable` describes what the method does better than `refresh`.

Reply:
How about `invalidateTableCache`? I think that would be more straightforward.

Reply:
That is a misleading name. This invalidates a single table's cached data, not the entire cache.

Reply:
The method accepts one parameter of Identifier type, which should not be misleading.
Just suggestion, overall I don't have a strong opinion on this.

Reply:
One quick question. This sounds like only to invalidate cached table metadata, and not touch query cache. But `Catalog.refreshTable` that is used v1 `REFRESH TABLE` command also drops `InMemoryRelation` if the table is cached in Spark query cache.

Is the difference intentional?

Reply:
table cache is managed by Spark, so the API here can not affect table cache inside Spark.

Reply:
@cloud-fan I think even though the table cached is managed by Spark, currently the behavior is different between v1 and v2 (which doesn't invalidate the table cache). Although it seems the whole table cache mechanism is largely implemented for v2, cached temp view can still be created via `CACHE TABLE ... AS SELECT` from a v2 table. Consequently, the cache which won't get invalidated when ppl call the `REFRESH TABLE` command on the v2 table.

Perhaps we should disallow caching a v2 table this way for now, and re-enable it once we have complete support for caching in v2.

Reply:
Yes, we should add this kind of missing actions into v2 commands as well. I think DROP TABLE has the same issue.